sdks/java/core/src/test/java/org/apache/beam/sdk/util/ReshuffleTest.java

```diff
@@ -88,6 +100,58 @@ public void testJustReshuffle() {
     pipeline.run();
   }
 
+  /**
+   * Tests that timestamps are not moved forwards in time after applying a {@link Reshuffle} with
+   * the default {@link WindowingStrategy}.
+   */
+  @Test
+  @Category(RunnableOnService.class)
+  public void testReshuffleDoesNotMoveTimestampsForwards() {
+    PCollection<KV<String, Instant>> input =
+        pipeline
+            .apply(
+                Create.timestamped(
+                        TimestampedValue.of("foo", BoundedWindow.TIMESTAMP_MIN_VALUE),
+                        TimestampedValue.of("foo", new Instant(0)),
+                        TimestampedValue.of("bar", new Instant(33)),
+                        TimestampedValue.of("bar", BoundedWindow.TIMESTAMP_MAX_VALUE))
+                    .withCoder(StringUtf8Coder.of()))
+            .apply("PairWithOriginalTimestamp", ParDo.of(new PairWithTimestampFn<String>()))
+            .setCoder(KvCoder.of(StringUtf8Coder.of(), InstantCoder.of()));
+
+    PCollection<KV<KV<String, Instant>, Instant>> output =
+        input
+            .apply(Reshuffle.<String, Instant>of())
+            .apply(
+                "PairWithReshuffledTimestamp",
+                ParDo.of(new PairWithTimestampFn<KV<String, Instant>>()));
+
+    PAssert.that(output)
+        .satisfies(
+            new SerializableFunction<Iterable<KV<KV<String, Instant>, Instant>>, Void>() {
+              @Override
+              public Void apply(Iterable<KV<KV<String, Instant>, Instant>> input) {
+                for (KV<KV<String, Instant>, Instant> elem : input) {
+                  Instant originalTimestamp = elem.getKey().getValue();
+                  Instant afterReshuffleTimestamp = elem.getValue();
+                  assertThat(
```

Comment:
Done. Because the underlying implementation is GroupByKey, we can't maintain the original timestamp without reifying it, but added a comment why the assert is the way it is.